pkg/ruler/lifecycle_test.go

```diff
@@ -2,73 +2,102 @@ package ruler
 
 import (
 	"context"
+	"sort"
 	"testing"
 	"time"
 
-	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
 	"github.com/cortexproject/cortex/pkg/ring"
+	"github.com/cortexproject/cortex/pkg/ring/kv/consul"
 	"github.com/cortexproject/cortex/pkg/ring/testutils"
 	"github.com/cortexproject/cortex/pkg/util/services"
 	"github.com/cortexproject/cortex/pkg/util/test"
 )
 
 // TestRulerShutdown tests shutting down ruler unregisters correctly
 func TestRulerShutdown(t *testing.T) {
+	ctx := context.Background()
+
 	config, cleanup := defaultRulerConfig(newMockRuleStore(mockRules))
-	config.EnableSharding = true
-	config.Ring.SkipUnregister = false
 	defer cleanup()
 
-	r, rcleanup := newTestRuler(t, config)
+	r, rcleanup := newRuler(t, config)
 	defer rcleanup()
 
+	r.cfg.EnableSharding = true
+	ringStore := consul.NewInMemoryClient(ring.GetCodec())
+
+	err := enableSharding(r, ringStore)
+	require.NoError(t, err)
+
+	require.NoError(t, services.StartAndAwaitRunning(ctx, r))
+	defer services.StopAndAwaitTerminated(ctx, r) //nolint:errcheck
+
 	// Wait until the tokens are registered in the ring
 	test.Poll(t, 100*time.Millisecond, config.Ring.NumTokens, func() interface{} {
-		return testutils.NumTokens(config.Ring.KVStore.Mock, "localhost", ring.RulerRingKey)
+		return testutils.NumTokens(ringStore, "localhost", ring.RulerRingKey)
 	})
 
 	require.NoError(t, services.StopAndAwaitTerminated(context.Background(), r))
```

Comment:
Could you also test the instance is `ACTIVE` in the ring?

Reply:
I'll update the test to do so